### PR TITLE
Android - Add a ReactFragment

### PR DESCRIPTION
**Pull request by jpshelley**
## Overview

React Native on Android has currently been focused and targeted at using an [Activity](https://developer.android.com/reference/android/app/Activity.html) for its main form of instantiation.

While this has probably worked for most companies and developers, you lose some of the modularity of a more cohesive application when working in a "brown-field" project that is currently native. This hurts more companies that are looking to adopt React Native and slowly implement it in a fully native application.

A lot of developers follow Android's guidelines of using Fragments in their projects, even if it is a debated subject in the Android community, and this addition will allow others to embrace React Native more freely. (I even assume it could help with managing navigation state in applications that contain a decent amount of Native code and would be appreciated in those projects. Such as sharing the Toolbar, TabBar, ViewPager, etc in Native Android) 

Even with this addition, a developer will still need to host the fragment in an activity, but now that activity can contain native logic like a Drawer, Tabs, ViewPager, etc.

**Test plan (required)**
* We have been using this class at Hudl for over a couple of months and have found it valuable.
* If the community agrees on the addition, I can add documentation to the Android sections to include notes about the potential of this Fragment.
* If the community agrees on the addition, I can update one or more of the examples in the `/Examples` folder and make use of the Fragment, or even create a new example that uses a native layout manager like Drawer, Tabs, Viewpager, etc)

Make sure tests pass on both Travis and Circle CI.

_To Note:_
* There is also talk of using React Native inside Android Fragment's without any legit documentation, this could help remedy some of that with more documentation included in this PR https://facebook.github.io/react-native/releases/0.26/docs/embedded-app-android.html#sharing-a-reactinstance-across-multiple-activities-fragments-in-your-app
* Others have also requested something similar and have a half-baked solution as well http://stackoverflow.com/questions/35221447/react-native-inside-a-fragment

## Release Notes
[ANDROID][FEATURE][ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java] - Adds support for Android's Fragment system. This allows for a more hybrid application.

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java [hunk 1]
```diff
@@ -0,0 +1,289 @@
+package com.facebook.react;
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java]: Add a copyright header please.

File: ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+import com.facebook.infer.annotation.Assertions;
+import com.facebook.react.common.ReactConstants;
+import com.facebook.react.devsupport.DoubleTapReloadRecognizer;
+import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler;
+import com.facebook.react.modules.core.PermissionAwareActivity;
+import com.facebook.react.modules.core.PermissionListener;
+
+public class ReactFragment extends Fragment implements PermissionAwareActivity, DefaultHardwareBackBtnHandler {
+
+  public final static int REQUEST_OVERLAY_CODE = 1111;
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java]: Why is this needed? Can you add a comment please?

File: ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+  private static final String REDBOX_PERMISSION_MESSAGE =
+    "Overlay permissions need to be granted in order for react native apps to run in dev mode.";
+  private static final String REDBOX_PERMISSION_GRANTED_MESSAGE =
+    "Overlay permissions have been granted.";
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java]: Why are these needed? Can you add a comment?

File: ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    }
+    if (getReactNativeHost().hasInstance()) {
+      getReactNativeHost().getReactInstanceManager().onHostDestroy(getActivity());
+    }
+  }
+  // endregion
+
+  /**
+   * This currently only checks to see if we've enabled the permission to draw over other apps.
+   * This is only used in debug/developer mode and is otherwise not used.
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java]: Used for what? Redbox? Please add a comment explaining why it's needed.

File: ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+      mReactRootView = null;
+    }
+    if (getReactNativeHost().hasInstance()) {
+      getReactNativeHost().getReactInstanceManager().onHostDestroy(getActivity());
+    }
+  }
+  // endregion
+
+  /**
+   * This currently only checks to see if we've enabled the permission to draw over other apps.
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java]: But why is this method needed at all? Why receive a result from an activity? What activity? The parent?

File: ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+        getReactNativeHost().getReactInstanceManager().showDevOptionsDialog();
+        handled = true;
+      }
+      boolean didDoubleTapR = Assertions.assertNotNull(mDoubleTapReloadRecognizer).didDoubleTapR(keyCode, getActivity().getCurrentFocus());
+      if (didDoubleTapR) {
+        getReactNativeHost().getReactInstanceManager().getDevSupportManager().handleReloadJS();
+        handled = true;
+      }
+    }
+    return handled;
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java]: This looks like copy & paste? Can you extract it to a shared helper rather than copy & pasting?

**Comment by mkonicek** [comment]
The overall idea seems reasonable.

However, how much of the code here is copied from somewhere else in the codebase? It's going to be very hard to maintain the exact same copy of the logic in two different places - it's guaranteed people will only update one place and forget to update the other. If we want to commit this every non-trivial piece of logic that's been copied from somewhere will need to be extracted to a shared helper.

Looks like the code doesn't compile?
https://circleci.com/gh/facebook/react-native/15316?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by jpshelley** [comment]
@mkonicek great insight and thanks for reviewing the code!
* Code doesn't compile because I used android annotations (nullable, non-null, etc) I can remove those and should compile
* I agree with the duplication, I think it hints at the code review comments you left too. This is basically a duplicate of the [ReactActivityDelegate](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java#L34) for a Fragment, I can attempt to share more of the code between the two. HOWEVER, a Fragment will always depend on its Activity, and therefor will need some of the callbacks and maybe an updated example or documentation will help?

File: ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java (hunk 2 above)
Comment (jpshelley) [inline ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java]: Yeah, its basically a duplication of the [ReactActivityDelegate] (https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java#L34) which relies on the permission system to allow dev/debug mode.

File: ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java (hunk 3 above)
Comment (jpshelley) [inline ReactAndroid/src/main/java/com/facebook/react/ReactFragment.java]: See comment above ^^

**Comment by jpshelley** [comment]
@mkonicek I've added a commit to fix this build failures, as well as attempt to re-use the existing code from the `ReactActivityDelegate`. I didn't want to eliminate the Delegate completely, so I created a new root delegate that `ReactActivityDelegate` calls into. I'm not sold on the current implementation but wanted to get something pushed so we could communicate potential changes and improvements.

I foresee something like the following:
react-delegate 1

The benefit of the Activity is less native code and easier to get a "green-field" project started.
The benefit of the Fragment is easier to integrate with an existing "brown-field" project and make use of native ViewGroups/Activities.

File: ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+  public ReactDelegate(Activity activity, String appKey, Bundle launchOptions) {
+    mActivity = activity;
+    mMainComponentName = appKey;
+    mLaunchOptions = launchOptions;
+    mDoubleTapReloadRecognizer = new DoubleTapReloadRecognizer();
+  }
+
+  public void onHostResume() {
+    if (getReactNativeHost().hasInstance()) {
+      if (mActivity instanceof DefaultHardwareBackBtnHandler) {
```
Comment (serj-lotutovici) [inline ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java]: This forces the host activity to implement the interface, when in fact it's already implemented by the `ReactFragment` (that btw is never used).

Comment (jpshelley) [inline ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java]: Good call out. How would you see this working then? Extra param in the constructor to pass a `DefaultHardwareBackBtnHandler` object which can either be a fragment or an activity that implements the interface?

Comment (serj-lotutovici) [inline ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java]: That would be my first choice. I wonder if the the behaviour would be the same, since `ReactActivity` is calling `super.onBackPressed()` and the fragment only calls `getActivity().onBackPressed()`. The chain of thought is that the host activity will have to call the fragment's `onBackPressed` method do decide whether or not to call the super method.

Comment (jpshelley) [inline ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java]: I'm actually implementing react-navigation this week in a project that will make of the back button in this exact fragment class so I can tweak it a bit to figure out what feels best. But I think I agree with what you're saying above.

Comment (jpshelley) [inline ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java]: @serj-lotutovici So I updated the code a bit in https://github.com/facebook/react-native/pull/12199/commits/b089982ebe9cec0e86a71bf4986c783ac5355e37

After messing with fragment navigations stacks this week, it seemed easier to allow the Activity to handle the back navigation. If the Fragment handled the back button handler, it ended up getting stuck in a loop, or it wasn't able to safely call the `super.onBackPressed()` like I was hoping for. I'm open to suggestions but this is the conclusion I came to.

This will be shown via documentation for sure, but I'll work on documentation if/when we get this PR "approved" and the idea is solidified.

**Comment by jpshelley** [comment]
I've ironed out more of the `ReactFragment` this week, however I'm wondering if it is worth refactoring [ReactActivityDelegate](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java#L34) now while RN is still in a `0.X.Y` stage and is able for "breaking changes"

The change I mention is similar to the idea I showed above via the diagram. The delegate will handle all common operations between a `ReactFragment` and `ReactActivity` such as dev tools, loading the app, and general interactions.

Let me know what people think about:
* The PR as a whole currently
* If the consensus is positive towards the question above, then where should the documentation live and should samples be updated to reflect a ReactFragment usage?
* Finally thoughts on a more shared delegate across the two.

**Comment by jpshelley** [comment]
Any thoughts on the build errors? Looks like a lot Flow related issue of `SectionList` which I believe was a recent commit.

File: ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java (hunk 7 above)
Comment (dobrienlzd) [inline ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java]: @jpshelley I agree on the importance of Fragment support.  For example some components like ViewPager return Fragments for each view.  This is standard.   Btw, are there tests for the Fragment solution or do you have an example app that is using these modifications?

Comment (jpshelley) [inline ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java]: We are currently in the process of open sourcing our solution at Hudl. Since contributing back to React Native seems to be taking longer then expected, we wanted to be able to share our approach with others. I will comment back on this PR when that happens.

**Comment by dobrienlzd** [comment]
Is there any example code you can push to github or bitbucket using the ReactFragment? This would help in it's use and review.

File: ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java (hunk 7 above)
Comment (dobrienlzd) [inline ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java]: Sounds great! I am currently doing React Native component integration into Native app and would like to try it out when available.

**Comment by githubdoramon** [comment]
@jpshelley Any chance you can show a example code for using with a view pager?

I tried the solution at http://stackoverflow.com/questions/35221447/react-native-inside-a-fragment, but the fragment don`t detect touches to it, so I can't perform any actions with it.

Thank you

**Comment by jpshelley** [comment]
For anyone following this PR, we are looking at open sourcing our solution at Hudl. From there we can add examples on how to include with ViewPagers or other Fragment based solutions.

**Comment by inakianduaga** [comment]
@jpshelley can you briefly list the instructions to get this working without waiting for a merge into the RN codebase (so moving all three classes into a separate namespace and adding it). I'm trying to get RN implemented in my current company and need to integrate it with an existing android app that uses fragments

**Comment by dobrienlzd** [comment]
Or provide an update @jpshelley on where you are at with the open sourcing
of the solution. Thanks

**Comment by jpshelley** [comment]
We're currently working on getting the correct OSS LICENSE and CLA setup. Its still in progress, but might be another two weeks unfortunately.

**Comment by githubdoramon** [comment]
Everyone, I created a Gist with an example on how I was able to make it work. Let me know if it is understandable.

https://gist.github.com/githubdoramon/c40108c293ff370f96211a9adedf95cc

**Comment by jpshelley** [comment]
For all those still following this PR, we've officially open sourced our fragment solution! 🎉 

You can check out the blog located here: https://medium.com/in-the-hudl/introducing-react-native-android-fragment-3559fe999ced

The project is located here: https://github.com/hudl/react-native-android-fragment

You can download it via jitpack. There is a sample app for more details. For those looking for a ViewPager solution I hope this helps. If not submit an issue or pull request to get it working in the sample application.

_I plan to keep this PR remained open for now though, until facebook/react-native officially says they won't support Fragments._ I feel it would be better supported upstream in the main project, but without knowing when that will happen, we will attempt to facilitate releases of React Native where possible.

**Comment by littlehome-eugene** [comment]
I think Facebook should state that fragements are not currently well supported in their https://facebook.github.io/react-native/docs/integration-with-existing-apps.html#integration-with-existing-apps doc.

**Comment by jpshelley** [comment]
@foghina @cmcewen @kmagiera if any of you are still working on React Native could we get another review of this PR? Its my understanding that Facebook internally doesn't use Fragments but a large deal of companies do and they would benefit from this.

**Comment by jpshelley** [comment]
@mdvacca @ayc1 @kmagiera I've noticed all of you have added to the android codebase, any chance we can get another review for this PR? I'm concerned as to why its been stale for so long.

File: ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    }
+  }
+
+  public void onHostPause() {
+    if (getReactNativeHost().hasInstance()) {
+      getReactNativeHost().getReactInstanceManager().onHostPause(mActivity);
+    }
+  }
+
+  public void onHostDetroy() {
```
Comment (jd009) [inline ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java]: Misspelling of 'Destroy'.

**Comment by vinceyuan** [comment]
+1
@mdvacca @ayc1 @kmagiera @mkonicek @serj-lotutovici @jd009 @dobrienlzd please review again.

File: ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java [hunk 9]
```diff
@@ -111,44 +105,19 @@ protected void onResume() {
...
 
   protected void onDestroy() {
-    if (mReactRootView != null) {
-      mReactRootView.unmountReactApplication();
-      mReactRootView = null;
-    }
-    if (getReactNativeHost().hasInstance()) {
-      getReactNativeHost().getReactInstanceManager().onHostDestroy(getPlainActivity());
-    }
+    mReactDelegate.onHostDetroy();
```
Comment (jd009) [inline ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java]: Minor misspelling here of "Destroy" --- should be `mReactDelegate.onHostDestroy();`

**Comment by jd009** [comment]
@vinceyuan Thanks for the heads up... but I do still see the misspelling of the word "Destroy".  Do you have a more recent version of the code which hasn't been pushed yet?  Thanks.

**Comment by vinceyuan** [comment]
@jpshelley Please respond.

**Comment by jpshelley** [comment]
I can fix that tomorrow, but is that the only thing stopping this feature from getting approved? It's a decent chunk of responsible for Facebook to maintain so I want to make sure the intention and architecture are okay, or is there other changes required? @jd009

**Comment by AyushG3112** [comment]
Is this blocked?

**Comment by hramos** [comment]
@dulmandakh you've sent some related PRs - do you have any feedback on this PR? It's been open for a while, and I could use some help deciding how to move forward.

**Comment by jpshelley** [comment]
Its worth mentioning, Android's transition to `androidx` and the `androidx` Fragment is taking priority over the Platform Fragments. --> https://developer.android.com/reference/androidx/fragment/app/Fragment

With that being said, I think my approach uses the support v4 Fragment, which is also considered "deprecated" in a sense. In order to support Android moving forward, the androidx libraries should be leveraged instead.

**Comment by hramos** [comment]
Alright, AndroidX is in the works (cc @mdvacca).

**Comment by dulmandakh** [comment]
IIRC, this PR is trying to add fragment with React Native support that can be included in native activities, or be used in parts of screen.

**Comment by cpojer** [comment]
This PR is more than two years old. I'm concerned at this stage about merging it now and instantly introducing technical debt in the form of "legacy" code. Is there anybody who would like to figure out a path forward for this change? What would it look like if this PR was created today, with Android X in mind?

**Comment by jpshelley** [comment]
@cpojer - honestly, It would be worth knowing what the "major rewrite" is going to be with React Native. Through blog posts, podcasts, etc the community has heard of this big change in React Native https://github.com/react-native-community/discussions-and-proposals/issues/4

I don't have enough time in my day-to-day to understand the rewrite personally, and I'm relying on FB engineers to take into consideration Android Fragments when doing the rewrite.

If that above doesn't happen, I would much prefer for the rewrite to be released and then create a new PR Introducing Android Fragments into that framework for Fabric.

**Comment by mdvacca** [comment]
Hey @jpshelley, thanks for working on this

The rewrite described in https://github.com/react-native-community/discussions-and-proposals/issues/4 will not affect these classes. 

I believe it is valuable to provide this functionality for OSS projects. If you rebase and make sure everything works I will review this PR

In terms of AndroidX, we are already working on the migration to AndroidX, since you are using support v4 Fragment that shouldn't be a problem here (it will be covered by the migration)

**Comment by cpojer** [comment]
@jpshelley Hey! Do you mind rebasing and testing your PR again so that @mdvacca can review and finally land it? :)

**Comment by jpshelley** [comment]
@cpojer @mdvacca you got it. I'll get started on it this week. Not sure how long it'll take to rebase 🤞

**Comment by jpshelley** [comment]
@cpojer @mdvacca Updated with master. Awesome to see the androidx support came in pretty early as well!

File: ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java [hunk 10]
```diff
@@ -195,12 +152,9 @@ public void onRequestPermissionsResult(
     final int requestCode,
     final String[] permissions,
     final int[] grantResults) {
-    mPermissionsCallback = new Callback() {
-      @Override
-      public void invoke(Object... args) {
-        if (mPermissionListener != null && mPermissionListener.onRequestPermissionsResult(requestCode, permissions, grantResults)) {
-          mPermissionListener = null;
-        }
+    mPermissionsCallback = args -> {
```
Comment (cpojer) [inline ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java]: Could you change this to not use a lambda expression? We are still supporting older versions of Java that don't have them.

**Comment by cpojer** [comment]
@jpshelley that's great! @mdvacca will do a review soon and we can hopefully get this landed finally. In the meantime, can you get the android tests to pass on CI? One of the issues is that you are using lambda expressions which unfortunately we cannot use yet.

File: ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java (hunk 10 above)
Comment (jpshelley) [reply ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java]: Yup definitely. Curious question, what version of Java do you back support and is there a specific reason why? Just wanting to learn from it.

File: ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java [hunk 11]
```diff
@@ -15,8 +15,6 @@
 
 import com.facebook.infer.annotation.Assertions;
 import com.facebook.react.bridge.Callback;
```
Comment (jpshelley) [inline ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java]: Before looking to merge this - does it make sense to have a `ReactActivityDelegate` and then a `ReactDelegate` or does it make sense for `ReactActivity` to use the `ReactDelegate` directly? Was trying to bring smaller changes with this PR and focus primarily on the addition of the `ReactFragment` but also don't want to incur possible Tech Debt

**Comment by cpojer** [comment]
OMG thank you @mdvacca for merging this PR. It's been open for over *two years* and I'm glad we could finally bring this functionality into React Native, especially since a ton of people have asked for it.

@jpshelley I also want to thank you for sticking with us for all this time and going through numerous reviews and rebases. I cannot imagine myself how demoralizing it must have been to wait for this long but I am incredibly thankful that you did. Since you sent this PR, the React Native team at Facebook changed completely and we try to spend more time on open source to bring your contributions into the repo faster. I hope to see many more PRs from you in the future, and I promise we'll get to them much faster.

Finally, thanks @mkonicek for reviewing this PR a long time ago and helping so much on open source at FB! I know how much work it was and that it's easy to drown in it.

**Comment by mdvacca** [comment]
We reverted this commit because it caused a bug internally at Facebook. I will work on a fix and bring the code back in a commit soon.

**Comment by jpshelley** [comment]
Anything to help make the environment and framework better @cpojer 🤝 

@mdvacca - let me know if you need any help at all.

and finally, is there any plans to update documentation/examples/blog posts etc? Would be happy to help on that front as well.

As soon as Hudl makes use of the merged code in a future release, we'll write up our own blog post to sunset https://github.com/hudl/react-native-android-fragment as well, with a migration plan in place.

**Comment by patrickkempff** [comment]
@mdvacca any progress on the fix? Would love to help if needed.

**Comment by mdvacca** [comment]
@patrickkempff this requires some changes internally at FB, I didn't have the time to take a look into it yet but I'm planning to work on it the next few weeks I will let you know If I need an extra hand!
Thanks

**Comment by mdvacca** [comment]
Today we fixed and landed this PR again.

**Comment by patrickkempff** [comment]
@mdvacca great news, awesome!

**Comment by saltyskip** [comment]
Any tutorial/guide on how to use this functionality?

**Comment by jpshelley** [comment]
@saltyskip until its released you can also leverage this library - https://github.com/hudl/react-native-android-fragment

this was the original work that got merged into master here. The API/Implementation should remain the same, so you can check out the example app.

Once Hudl is able to use the React-Native version, I'll follow-up and add documentation to React Native and an example, if someone else doesn't beat me to it.

**Comment by saltyskip** [comment]
Thanks @jpshelley got it to work with your original library. Great work, and looking forward to when this functionality is included in an official release :)

**Comment by devvit** [comment]
```ReactDelegate``` should be exposed, consider the scene [react-navigation](https://reactnavigation.org/docs/en/getting-started.html)

```java
public class MainActivity extends ReactActivity {

  @Override
  protected String getMainComponentName() {
    return "Example";
  }

+  @Override
+  protected ReactActivityDelegate createReactActivityDelegate() {
+    return new ReactActivityDelegate(this, getMainComponentName()) {
+      @Override
+      protected ReactRootView createRootView() {
+       return new RNGestureHandlerEnabledRootView(MainActivity.this);
+      }
+    };
+  }
}
```

**Comment by IanVS** [comment]
I found my way to this PR after trying to add react-navigation to an old brownfield app that's slowly been migrated over the years up to RN 0.59.  It seems to use the strategy in https://facebook.github.io/react-native/docs/integration-with-existing-apps, but those docs say nothing about `ReactActivity` or `MainActivity`, so I'm not sure what I need to do to get `react-native-gesture-handler` working as @devvit shows above.  Are the docs still correct, in that I should be extending my activity from `Activity`?  Or should I now extend from `ReactActivity` which I guess was added by this PR?  Sorry to spam this old PR, but I'm not a native android developer and I've been struggling with this most of the day.  My main question is whether https://facebook.github.io/react-native/docs/integration-with-existing-apps is up-to-date, or if there are any better resources for how to integrate RN into an existing android app.  🙏 🙇

**Comment by dulmandakh** [comment]
If you want to use React, then ReactActivity is the way to go.

**Comment by IanVS** [comment]
That's great to hear, thanks.  Is there any documentation about it, since apparently the official docs are out-of-date?

**Comment by BharathMG-ShareChat** [comment]
> @mdvacca `ReactDelegate` should be exposed, consider the scene [react-navigation](https://reactnavigation.org/docs/en/getting-started.html)
> 
> ```java
> public class MainActivity extends ReactActivity {
> 
>   @Override
>   protected String getMainComponentName() {
>     return "Example";
>   }
> 
> +  @Override
> +  protected ReactActivityDelegate createReactActivityDelegate() {
> +    return new ReactActivityDelegate(this, getMainComponentName()) {
> +      @Override
> +      protected ReactRootView createRootView() {
> +       return new RNGestureHandlerEnabledRootView(MainActivity.this);
> +      }
> +    };
> +  }
> }
> ```
> 
> Currently(RN 0.61), there isnt a interface to do that

How to solve this issue? Currently, only way is to rewrite `ReactFragment` and wrap `ReactRootView` with Gesture handler?

**Comment by mitsest** [comment]
Hello,
so I 'm trying to use com.facebook.react.ReactFragment by extending it. 

Unfortunately, that's not possible 
```
private static ReactFragment newInstance
```

This should be
```
public static Bundle getReactFragmentBundle()
````

This way I can get a bundle that's required by your ReactFragment, and add my own args in it.
The only way I can use your solution is by copypasting the entire ReactFragment code.

Any plans on fixing this?

**Comment by keeslinp** [comment]
For any future folks with @mitsest's question. You can use the `Builder` to do something like `ReactFragment.Builder().setComponentName(name).setLaunchArguments(launchOptionsBundle)` instead of using `newInstance` directly
